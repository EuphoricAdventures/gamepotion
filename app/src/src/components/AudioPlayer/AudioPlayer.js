import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import icons from '../../icons'

import Button from '../Button/Button'

const StyledAudioPlayer = styled.div`
  // background-color: yellow;
  .component--button {
    display: inline-block;
    width: 4rem;
  }
  .component--button:not(:last-of-type) {
    margin-right: 0.5rem;
  }
`

class AudioPlayer extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      url: props.url,
      isPlaying: false,
      disabled: false
    }
    this.audio = new window.Audio()
    this.audio.onended = () => {
      this.stop()
    }
    this.audio.onerror = () => {
      this.setState({
        isPlaying: false,
        disabled: true
      })
    }
    this.playStop = this.playStop.bind(this)
  }

  componentDidMount() {
    this.load(this.state.url)
  }

  componentWillReceiveProps (nextProps) {
    this.setState({ url: nextProps.url })
    this.load(nextProps.url)
  }

  load(url) {
    if (typeof url === 'string') {
      this.audio.src = url
      this.audio.load()
      this.setState({
        disabled: false
      })
    } else {
      this.setState({
        disabled: true
      })
    }
  }

  play() {
    const mightBeAPromise = this.audio.play()
    if (mightBeAPromise !== undefined) {
      mightBeAPromise
        .then(() => {
          this.setState({
            isPlaying: true
          })
        })
        .catch(() => {
          this.setState({
            isPlaying: false
          })
        })
    } else {
      this.setState({
        isPlaying: true
      })
    }
  }

  stop() {
    this.audio.currentTime = 0
    this.audio.pause()
    this.setState({
      isPlaying: false
    })
  }

  playStop() {
    if (this.state.isPlaying === true) {
      this.stop()
    } else {
      this.play()
    }
  }

  render() {
    const buttonIcon = (this.state.isPlaying === false ? icons.generic.resource.sound.play : icons.generic.resource.sound.stop)
    return (
      <StyledAudioPlayer className='component--audio-player' ariaDisabled={this.state.disabled}>
        <Button icon={buttonIcon} onClick={this.playStop} disabled={this.state.disabled} />
      </StyledAudioPlayer>
    )
  }
}

AudioPlayer.propTypes = {
  url: PropTypes.oneOfType([
    PropTypes.string
  ]).isRequired
}

AudioPlayer.defaultProps = {
  url: null
}

export default AudioPlayer
