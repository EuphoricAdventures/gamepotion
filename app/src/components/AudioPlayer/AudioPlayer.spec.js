import React from 'react'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'
import 'jest-styled-components'
import AudioPlayer from './AudioPlayer'

window.HTMLMediaElement.prototype.load = jest.fn()

it('renders', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <AudioPlayer url='xyz' />
      </MemoryRouter>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
