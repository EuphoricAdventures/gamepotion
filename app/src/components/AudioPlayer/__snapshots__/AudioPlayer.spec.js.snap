// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c1 {
  display: block;
  padding: 0.7rem;
  -webkit-transition: background-color 0.2s ease-in-out,border-color 0.2s ease-in-out,opacity 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,border-color 0.2s ease-in-out,opacity 0.2s ease-in-out;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  background-color: #2e3131;
  color: #ffffff;
  border-radius: 4px;
  border-width: 2px;
  border-style: solid;
  border-color: #2e3131;
  outline: 0;
}

.c1[disabled] {
  opacity: 0.5;
}

.c1:focus {
  border-color: #38EF7D;
}

.c1:not([disabled]):hover {
  background-color: #6c7a89;
  border-color: #6c7a89;
}

.c1 > img {
  display: block;
  width: 2rem;
  height: 2rem;
  margin: 0 auto 0 auto;
}

.c1 > span {
  display: block;
  height: 1.5rem;
  line-height: 1.5rem;
}

.c1 > img + span {
  margin-top: 0.5rem;
}

.c0 .component--button {
  display: inline-block;
  width: 4rem;
}

.c0 .component--button:not(:last-of-type) {
  margin-right: 0.5rem;
}

<div
  className="component--audio-player c0"
>
  <button
    className="component--button c1"
    disabled={false}
    onClick={[Function]}
    title=""
  >
    <img
      alt=""
      src="resource-sound-play.svg"
    />
  </button>
</div>
`;
